File: week13/10937/10937_김도훈.py
import sys
sys.stdin = open("input.txt", 'r')

N = int(input())
arr = list(list(map(lambda c: ord(c) - ord('A'), input())) for _ in range(N))
dp = [[0] * (1<<N) for _ in range(N)]

cost = [[100, 70, 40, 0, 0, 0],
        [70, 50, 30, 0, 0, 0],
        [40, 30, 20, 0, 0, 0],
        [0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0]
        ]

# dp 초기화
# 0번째 행
for j in range(1, N):
    for curr_bit in range(1<<N):
        if (not (curr_bit & 1<<j)) and (not (curr_bit & 1<<(j-1))):
            dp[0][(curr_bit | 1<<j) | 1<<(j-1)] = max(dp[0][curr_bit | 1<<j], dp[0][curr_bit] + cost[arr[0][j]][arr[0][j-1]])



# i번째 행
for i in range(1, N):
    # j번째 열 킬지 말지
    prev_max = 0
    for prev_bit in range(1<<N):
        prev_max = max(prev_max, dp[i-1][prev_bit])
    dp[i][0] = prev_max

    for curr_bit in range(1<<N):
        # 킬 수 있는 경우 - 세로
        for prev_bit in range(1<<N):
            temp = dp[i-1][prev_bit]
            for j in range(N):
                if (curr_bit & 1<<j):
                    if (prev_bit & 1<<j):
                        break
                    else:
                        temp += cost[arr[i][j]][arr[i-1][j]]
            else:
                dp[i][curr_bit] = max(dp[i][curr_bit], temp)      
            

    # 킬 수 있는 경우 - 가로
    for curr_bit in range(1<<N):
        for j in range(1, N):
            if not (curr_bit & 1<<j) and not (curr_bit & 1<<(j-1)):
                new_bit = curr_bit | 1<<j | 1<<(j-1)
                dp[i][new_bit] = max(dp[i][new_bit], dp[i][curr_bit] + cost[arr[i][j]][arr[i][j-1]])
        

ans = 0
for curr_bit in range(1<<N):
    ans = max(ans, dp[N-1][curr_bit])


# for i in range(N):
#     print(i+1,"번째 dp:", end="")
#     for idx, elem in enumerate(dp[i]):
#         print([elem, format(idx, '05b')], end=", ")
#     print()
#     print("=================")

print(ans)

File: week13/10937/10937_김도훈.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    int N;
    cin >> N;

    // 문자를 숫자로 매핑하여 배열을 초기화
    vector<vector<int>> arr(N, vector<int>(N));
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            char c;
            cin >> c;
            arr[i][j] = c - 'A';
        }
    }

    // dp 배열을 2차원 벡터로 초기화
    vector<vector<int>> dp(N, vector<int>(1 << N, 0));

    // 비용 행렬 초기화
    vector<vector<int>> cost = {
        {100, 70, 40, 0, 0, 0},
        {70, 50, 30, 0, 0, 0},
        {40, 30, 20, 0, 0, 0},
        {0, 0, 0, 0, 0, 0},
        {0, 0, 0, 0, 0, 0},
        {0, 0, 0, 0, 0, 0}
    };

    // 0번째 행을 초기화
    for (int j = 1; j < N; j++) {
        for (int curr_bit = 0; curr_bit < (1 << N); curr_bit++) {
            if (!(curr_bit & (1 << j)) && !(curr_bit & (1 << (j - 1)))) {
                int new_bit = (curr_bit | (1 << j)) | (1 << (j - 1));
                dp[0][new_bit] = max(dp[0][new_bit], dp[0][curr_bit] + cost[arr[0][j]][arr[0][j - 1]]);
            }
        }
    }

    // i번째 행 계산
    for (int i = 1; i < N; i++) {
        int prev_max = *max_element(dp[i - 1].begin(), dp[i - 1].end());
        dp[i][0] = prev_max;

        // 세로 연결 처리
        for (int curr_bit = 0; curr_bit < (1 << N); curr_bit++) {
            for (int prev_bit = 0; prev_bit < (1 << N); prev_bit++) {
                int temp = dp[i - 1][prev_bit];
                bool valid = true;
                for (int j = 0; j < N; j++) {
                    if (curr_bit & (1 << j)) {
                        if (prev_bit & (1 << j)) {
                            valid = false;
                            break;
                        }
                        else {
                            temp += cost[arr[i][j]][arr[i - 1][j]];
                        }
                    }
                }
                if (valid) {
                    dp[i][curr_bit] = max(dp[i][curr_bit], temp);
                }
            }
        }

        // 가로 연결 처리
        for (int curr_bit = 0; curr_bit < (1 << N); curr_bit++) {
            for (int j = 1; j < N; j++) {
                if (!(curr_bit & (1 << j)) && !(curr_bit & (1 << (j - 1)))) {
                    int new_bit = curr_bit | (1 << j) | (1 << (j - 1));
                    dp[i][new_bit] = max(dp[i][new_bit], dp[i][curr_bit] + cost[arr[i][j]][arr[i][j - 1]]);
                }
            }
        }
    }

    // 최대 점수 계산
    int ans = *max_element(dp[N - 1].begin(), dp[N - 1].end());
    cout << ans << endl;

    return 0;
}
